mavros/src/plugins/global_position.cpp

```diff
@@ -168,12 +170,17 @@ class GlobalPositionPlugin : public plugin::PluginBase {
 		float eph = (raw_gps.eph != UINT16_MAX) ? raw_gps.eph / 1E2F : NAN;
 		float epv = (raw_gps.epv != UINT16_MAX) ? raw_gps.epv / 1E2F : NAN;
 
-		if (!std::isnan(eph)) {
-			const double hdop = eph;
-
-			// From nmea_navsat_driver
-			fix->position_covariance[0 + 0] = fix->position_covariance[3 + 1] = std::pow(hdop, 2);
-			fix->position_covariance[6 + 2] = std::pow(2 * hdop, 2);
+		// With FCU protocol v2.0 use accuracies reported by sensor
+		if (UAS_FCU(m_uas)->get_protocol_version() == mavconn::Protocol::V20 &&
+				raw_gps.h_acc > 0 && raw_gps.v_acc > 0) {
+			fix->position_covariance[0 + 0] = fix->position_covariance[3 + 1] = std::pow(raw_gps.h_acc / 1E3, 2);
+			fix->position_covariance[6 + 2] = std::pow(raw_gps.v_acc / 1E3, 2);
+			fix->position_covariance_type = sensor_msgs::NavSatFix::COVARIANCE_TYPE_DIAGONAL_KNOWN;
```

Comment:
About filling cov matrix, i'd like to see Eigen::Map here ([ftf::EigenMapCovariance3d][1]), old code done long before we start to use Eigen.

[1]: http://docs.ros.org/lunar/api/mavros/html/group__nodelib.html#gabafbfa3aa489d1c304f320a417ce91d1

Reply:
Pseudocode:
```cpp
ftf::EigenMapCovariance3d cov(fix->position_covariance.data());

xy = pow(...);
cov.diagonal() << xy, xy, z;
```